File: shared/playerState.js
// Player state for resource and supply tracking

// Get constants based on environment
const _getConstants = () => {
  if (typeof module !== 'undefined' && module.exports) {
    return require('./constants');
  }
  return window.CONSTANTS;
};

class PlayerState {
  constructor(playerId, playerIndex) {
    const CONSTANTS = _getConstants();

    this.playerId = playerId;
    this.playerIndex = playerIndex; // 0 or 1
    this.minerals = CONSTANTS.STARTING_MINERALS;
    this.supply = CONSTANTS.STARTING_SUPPLY;
    this.maxSupply = CONSTANTS.STARTING_MAX_SUPPLY;
  }

  canAfford(cost) {
    return this.minerals >= cost;
  }

  canSupport(supplyCost) {
    return this.supply + supplyCost <= this.maxSupply;
  }

  spend(amount) {
    if (!this.canAfford(amount)) return false;
    this.minerals -= amount;
    return true;
  }

  addMinerals(amount) {
    this.minerals += amount;
  }

  useSupply(amount) {
    this.supply += amount;
  }

  freeSupply(amount) {
    this.supply = Math.max(0, this.supply - amount);
  }

  addMaxSupply(amount) {
    this.maxSupply += amount;
  }

  toJSON() {
    return {
      playerId: this.playerId,
      playerIndex: this.playerIndex,
      minerals: this.minerals,
      supply: this.supply,
      maxSupply: this.maxSupply
    };
  }

  static fromJSON(data) {
    const state = new PlayerState(data.playerId, data.playerIndex);
    state.minerals = data.minerals;
    state.supply = data.supply;
    state.maxSupply = data.maxSupply;
    return state;
  }
}

// Export for both Node.js and browser
if (typeof module !== 'undefined' && module.exports) {
  module.exports = PlayerState;
} else if (typeof window !== 'undefined') {
  window.PlayerState = PlayerState;
}

File: server/Rooms/PlayerState.cs
using server.Setup;

namespace server.Rooms;

/// <summary>
/// Tracks player resources and supply during a game.
/// </summary>
public class PlayerState
{
    public string PlayerId { get; }
    public int PlayerIndex { get; }  // 0 or 1
    public int Minerals { get; private set; }
    public int Supply { get; private set; }
    public int MaxSupply { get; private set; }

    public PlayerState(string playerId, int playerIndex)
    {
        PlayerId = playerId;
        PlayerIndex = playerIndex;
        Minerals = GameConstants.STARTING_MINERALS;
        Supply = GameConstants.STARTING_SUPPLY;
        MaxSupply = GameConstants.STARTING_MAX_SUPPLY;
    }

    public bool CanAfford(int cost)
    {
        return Minerals >= cost;
    }

    public bool CanSupport(int supplyCost)
    {
        return Supply + supplyCost <= MaxSupply;
    }

    public bool Spend(int amount)
    {
        if (!CanAfford(amount))
            return false;
        Minerals -= amount;
        return true;
    }

    public void AddMinerals(int amount)
    {
        Minerals += amount;
    }

    public void UseSupply(int amount)
    {
        Supply += amount;
    }

    public void FreeSupply(int amount)
    {
        Supply = Math.Max(0, Supply - amount);
    }

    public void AddMaxSupply(int amount)
    {
        MaxSupply += amount;
    }

    public Dictionary<string, object> ToJson()
    {
        return new Dictionary<string, object>
        {
            ["playerId"] = PlayerId,
            ["playerIndex"] = PlayerIndex,
            ["minerals"] = Minerals,
            ["supply"] = Supply,
            ["maxSupply"] = MaxSupply
        };
    }
}
